flink-streaming-java/src/main/java/org/apache/flink/streaming/api/functions/TimeBoundedStreamJoinOperator.java

```diff
@@ -0,0 +1,328 @@
+	private static final String LAST_CLEANUP_RIGHT = "LAST_CLEANUP_RIGHT";
+
+	private transient ValueState<Long> lastCleanupRightBuffer;
+	private transient ValueState<Long> lastCleanupLeftBuffer;
+
+	private transient MapState<Long, List<Tuple3<T1, Long, Boolean>>> leftBuffer;
+	private transient MapState<Long, List<Tuple3<T2, Long, Boolean>>> rightBuffer;
+
+	private transient TimestampedCollector<Tuple2<T1, T2>> collector;
+
+	/**
+	 * Creates a new TimeBoundedStreamJoinOperator
+	 *
+	 * @param lowerBound          The lower bound for evaluating if elements should be joined
+	 * @param upperBound          The upper bound for evaluating if elements should be joined
+	 * @param lowerBoundInclusive Whether or not to include elements where the timestamp matches
+	 *                            the lower bound
+	 * @param upperBoundInclusive Whether or not to include elements where the timestamp matches
+	 *                            the upper bound
+	 */
+	public TimeBoundedStreamJoinOperator(long lowerBound,
+										 long upperBound,
+										 boolean lowerBoundInclusive,
+										 boolean upperBoundInclusive) {
+
+		this.lowerBound = lowerBound;
+		this.upperBound = upperBound;
+
+		this.inverseLowerBound = -1 * upperBound;
+		this.inverseUpperBound = -1 * lowerBound;
+
+		this.lowerBoundInclusive = lowerBoundInclusive;
+		this.upperBoundInclusive = upperBoundInclusive;
+	}
+
+	@Override
+	public void open() throws Exception {
+		super.open();
+		collector = new TimestampedCollector<>(output);
+
+		this.leftBuffer = getRuntimeContext().getMapState(new MapStateDescriptor<>(
+			LEFT_BUFFER,
+			LONG_TYPE_INFO,
+			TypeInformation.of(new TypeHint<List<Tuple3<T1, Long, Boolean>>>() {
+			})
+		));
+
+		this.rightBuffer = getRuntimeContext().getMapState(new MapStateDescriptor<>(
+			RIGHT_BUFFER,
+			LONG_TYPE_INFO,
+			TypeInformation.of(new TypeHint<List<Tuple3<T2, Long, Boolean>>>() {
+			})
+		));
+
+		this.lastCleanupRightBuffer = getRuntimeContext().getState(new ValueStateDescriptor<>(
+			LAST_CLEANUP_RIGHT,
+			LONG_TYPE_INFO
+		));
+
+		this.lastCleanupLeftBuffer = getRuntimeContext().getState(new ValueStateDescriptor<>(
+			LAST_CLEANUP_LEFT,
+			LONG_TYPE_INFO
+		));
+	}
+
+	/**
+	 * Process a {@link StreamRecord<T1>} from the left stream. Whenever an {@link StreamRecord<T1>}
+	 * arrives at the left stream, it will get added to the left buffer. Possible join candidates
+	 * for that element will be looked up from the right buffer and if the pair lies within the
+	 * user defined boundaries, it gets collected.
+	 *
+	 * @param record An incoming record to be joined
+	 * @throws Exception Can throw an Exception during state access
+	 */
+	@Override
+	public void processElement1(StreamRecord<T1> record) throws Exception {
+
+		long leftTs = record.getTimestamp();
+		T1 leftValue = record.getValue();
+
+		addToLeftBuffer(leftValue, leftTs);
+
+		getJoinCandidatesForLeftElement(leftTs)
```

Comment:
This code creates unnecessary copies, as it first copies everything from the state to a list and then iterates over the list checking which elements are able to be joined. Why not keeping the for-loop and writing sth like:

```
for (long i = lowerBound; i <= upperBound; i++) {
			List<Tuple3<T2, Long, Boolean>> fromBucket = rightBuffer.get(leftTs + i);
			if (fromBucket != null) {
				for (Tuple3<T2, Long, Boolean> candidate : fromBucket) {
					if (shouldBeJoined(leftTs, candidate.f1)) {
						collect(leftValue, candidate.f0, leftTs);
					}
				}
			}
		}
```